python/pylibcudf/pylibcudf/tests/conftest.py

```diff
@@ -83,80 +83,87 @@ def _get_vals_of_type(pa_type, length, seed):
+            def _generate_nested_data(typ):
+                child_colnames = []
+
+                # recurse to get vals for children
+                rand_arrs = []
+                for i in range(typ.num_fields):
+                    rand_arr, grandchild_colnames = _generate_nested_data(
+                        typ.field(i).type
+                    )
+                    rand_arrs.append(rand_arr)
+                    child_colnames.append(
+                        (typ.field(i).name, grandchild_colnames)
+                    )
+
+                if isinstance(typ, pa.StructType):
+                    pa_array = pa.StructArray.from_arrays(
+                        [rand_arr for rand_arr in rand_arrs],
+                        names=[
+                            typ.field(i).name for i in range(typ.num_fields)
+                        ],
+                    )
+                elif isinstance(typ, pa.ListType):
+                    pa_array = pa.array(
+                        [list(row_vals) for row_vals in zip(rand_arrs[0])],
+                        type=typ,
+                    )
+                    child_colnames.append(("", grandchild_colnames))
+                else:
+                    # typ is scalar type
+                    pa_array = pa.array(
+                        _get_vals_of_type(typ, nrows, seed=seed), type=typ
+                    )
+                return pa_array, child_colnames
+
+            if isinstance(typ, (pa.ListType, pa.StructType)):
+                rand_arr, child_colnames = _generate_nested_data(typ)
             else:
-                # typ is scalar type
-                pa_array = pa.array(
+                rand_arr = pa.array(
                     _get_vals_of_type(typ, nrows, seed=seed), type=typ
                 )
-            return pa_array, child_colnames
 
-        if isinstance(typ, (pa.ListType, pa.StructType)):
-            rand_arr, child_colnames = _generate_nested_data(typ)
-        else:
-            rand_arr = pa.array(
-                _get_vals_of_type(typ, nrows, seed=seed), type=typ
-            )
+            table_dict[f"col_{typ}"] = rand_arr
+            colnames.append((f"col_{typ}", child_colnames))
 
-        table_dict[f"col_{typ}"] = rand_arr
-        colnames.append((f"col_{typ}", child_colnames))
+        pa_table = pa.Table.from_pydict(table_dict)
 
-    pa_table = pa.Table.from_pydict(table_dict)
+        return plc.io.TableWithMetadata(
+            plc.interop.from_arrow(pa_table), column_names=colnames
+        ), pa_table
 
-    return plc.io.TableWithMetadata(
-        plc.interop.from_arrow(pa_table), column_names=colnames
-    ), pa_table
+    return _table_data
 
 
 @pytest.fixture(params=[(0, 0), ("half", 0), (-1, "half")])
 def nrows_skiprows(table_data, request):
     """
     Parametrized nrows fixture that accompanies table_data
     """
-    _, pa_table = table_data
+    _, pa_table = table_data()
```

Comment:
Curious why you needed to change to to return a callable instead of returning the data. This will disallow pytest from caching the generated data.

Reply:
`write_csv` fails when the the table has struct and list types. Is that true? The error I see is `RuntimeError: CUDF failure at:/home/coder/cudf/cpp/src/io/csv/writer_impl.cu:277: Unsupported column type.`

So I wanted a way to create the `table_data` fixture with a specific set of pyarrow types. So I tried to follow this [factory fixture pattern](https://www.inspiredpython.com/article/five-advanced-pytest-fixture-patterns) which returns a callable. This is why I'm calling it like `table_data(pa_types=NON_NESTED_PA_TYPES)` or `table_data() # default is DEFAULT_PA_TYPES`. The factory fixture has session scope, so I guess the function (not generated table) is cached and reused per session? I'm not sure what to do here.

Reply:
I would suggest just skipping the write csv test for those types instead since it seems like a libcudf limitation

Reply:
Technically it may be a limit to CSV. 
What would a struct or list look like in a comma-separated text file? 
I cannot imagine anything that does not just quickly turn into JSON.

Reply:
Fair, I don't think this is well defined in CSV as opposed to libcudf.

At least in pandas, it just returns a string representation of the nested structures

```python
In [2]: df = pd.DataFrame([[[1, "str", True]]])

In [3]: df
Out[3]: 
                0
0  [1, str, True]

In [4]: df.to_csv()
Out[4]: ',0\n0,"[1, \'str\', True]"\n'
```

Reply:
Ah, that is good to know. What does a `struct` column look like?

Reply:
Similar to the list case, a Python string representation of the "struct". pandas doesn't have a dedicated struct type so it's not necessarily valid JSON

```python
In [8]: df = pd.DataFrame([[{1: True, "a": {"b": 2}}]])

In [9]: df
Out[9]: 
                          0
0  {1: True, 'a': {'b': 2}}

In [10]: df.to_csv()
Out[10]: ',0\n0,"{1: True, \'a\': {\'b\': 2}}"\n'

In [14]: df = pd.DataFrame([[{"1": 2.0, "2": 3.5}]])

In [15]: df
Out[15]: 
                      0
0  {'1': 2.0, '2': 3.5}

In [16]: df.to_csv()
Out[16]: ',0\n0,"{\'1\': 2.0, \'2\': 3.5}"\n'
```